From f67471319b8afd47288d69e19146941847260a97 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Thu, 21 Nov 2024 17:13:36 +0100
Subject: [PATCH 31/82] phy: stm32: usb2phy: fix regulator enabling for wakeup

The regulators should remain enabled, when the USB controller
has kept the usb2phy initialized, e.g. for wakeup.
So add a new flag, do catch it at suspend time: when disabling the PHY
clocks (un-needed for wakeup), the regulators needs to remain active.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/linux.git d316555d39a628c257671dfd0e2091529fdf14c5 phy: stm32: usb2phy: fix regulator enabling for wakeup]
Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I4a8a249340094ca9fa75ee96478eafde69ec4794
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/420288
Reviewed-by: Ram DAYAL <ram.dayal@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Ram DAYAL <ram.dayal@st.com>
---
 drivers/phy/st/phy-stm32-usb2phy.c | 25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/drivers/phy/st/phy-stm32-usb2phy.c b/drivers/phy/st/phy-stm32-usb2phy.c
index 37a015872eff6..2d5c1adf0d0e5 100644
--- a/drivers/phy/st/phy-stm32-usb2phy.c
+++ b/drivers/phy/st/phy-stm32-usb2phy.c
@@ -86,6 +86,7 @@ struct stm32_usb2phy {
 	struct clk_hw clk48_hw;
 	atomic_t en_refcnt;
 	const struct stm32mp2_usb2phy_hw_data *hw_data;
+	bool do_wakeup;
 };
 
 enum stm32_usb2phy_mode {
@@ -270,10 +271,13 @@ static int stm32_usb2phy_enable(struct stm32_usb2phy *phy_dev)
 			phy_dev->value_trim2);
 	}
 
-	ret = stm32_usb2phy_regulators_enable(phy_dev);
-	if (ret) {
-		dev_err(dev, "can't enable regulators (%d)\n", ret);
-		return ret;
+	/* balance regulator_enable calls (kept ON during suspend, with wakeup enabled) */
+	if (!phy_dev->do_wakeup) {
+		ret = stm32_usb2phy_regulators_enable(phy_dev);
+		if (ret) {
+			dev_err(dev, "can't enable regulators (%d)\n", ret);
+			return ret;
+		}
 	}
 
 	ret = clk_prepare_enable(phy_dev->phyref);
@@ -314,10 +318,13 @@ static int stm32_usb2phy_disable(struct stm32_usb2phy *phy_dev)
 
 	clk_disable_unprepare(phy_dev->phyref);
 
-	ret = stm32_usb2phy_regulators_disable(phy_dev);
-	if (ret) {
-		dev_err(phy_dev->dev, "can't disable regulators (%d)\n", ret);
-		return ret;
+	if (!phy_dev->do_wakeup) {
+		/* Need to keep regulators ON for the wakeup case */
+		ret = stm32_usb2phy_regulators_disable(phy_dev);
+		if (ret) {
+			dev_err(phy_dev->dev, "can't disable regulators (%d)\n", ret);
+			return ret;
+		}
 	}
 
 	return 0;
@@ -334,6 +341,7 @@ static int stm32_usb2phy_suspend(struct device *dev)
 	 * usb2-phy is already turned off by HCD driver using exit callback
 	 */
 	if (phy_dev->is_init) {
+		phy_dev->do_wakeup = true;
 		ret = stm32_usb2phy_disable(phy_dev);
 		if (ret) {
 			dev_err(dev, "can't disable usb2phy (%d)\n", ret);
@@ -360,6 +368,7 @@ static int stm32_usb2phy_resume(struct device *dev)
 			dev_err(dev, "can't enable usb2phy (%d)\n", ret);
 			return ret;
 		}
+		phy_dev->do_wakeup = false;
 	}
 
 	return 0;
-- 
2.51.0

